### PR TITLE
General histogram collection

### PR DESCRIPTION
**Pull request by benkrikler**
This PR takes over from #31, using a common branch on the official repo that Luke and I can contribute to.  

The original message from #31:

> Added first tests for new histogram collection.
> 
> Some of the functionality is still unclear, e.g. how to introduce generators for different dimensions and how they work together.
> E.g. a pileup bins generator would create one histogram (with different names) for each pileup bin when
> ```python
> hists.add('myHist', bins=[0,100,200,450])
> ```
> is called.
> 
> However, when you bin in both detector regions AND pileup, then only one of the generators should actually create the histogram object, while the other one would only add the binning to the histogram name.

### CODE REVIEW

File: cmsl1t/collections/hist.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        '''
+        index_list = self._find_bins(keys)
+        for indices in index_list:
+            self._hists[indices] = value
+
+    def shape(self):
+        _shape = [len(dim) for dim in self._dimensions]
+        return tuple(_shape)
+
+    def __len__(self):
```
Comment (benkrikler) [inline cmsl1t/collections/hist.py]: I'm a little unsure what we expect `len` to mean on a histogram collection. I can think of a few different meanings: the number of currently filled histograms, the total number of key combinations (= histograms that could be filled), the total number of dimensions, or the number of bins in the first dimension.  Given the potential ambiguity, I'm tempted to remove this method altogether.

Comment (kreczko) [inline cmsl1t/collections/hist.py]: The number of histograms it contains. This one is going towards the iterators and such, a good exercise. For the number of combinations, you can have a separate function (e.g  `volume`?)

**Comment by benkrikler** [comment]
@kreczko, I think the histogram collection class is finished now. The creation of the actual, underlying histograms is delegated to a histogram_factory class that I will now create, but basically this will wrap the original ROOT histograms (or call the correct rootpy methods if available).  

I'm tempted to rearrange the histograms a bit. Since there will (once this is finished) only be one collections class, I might rename the directory just `hists` and include in it the base and derived histogram classes and the factory class as well.  How does that sound?

Once these are all finished, the next big task is to look at the plotting aspects...

**Comment by kreczko** [comment]
> I'm tempted to rearrange the histograms a bit. Since there will (once this is finished) only be one collections class, I might rename the directory just hists and include in it the base and derived histogram classes and the factory class as well. How does that sound?

Sounds good to me. We can always go back to `collections` if we get another use case.

**Comment by benkrikler** [comment]
@kreczko, I think the collection stuff is finished now. To create a collection of histograms, you need to set up the dimensions for the collection's binning and then create a collection instance by also specifying a histogram factory.  You can either roll your own factory (which is just a method that should return a histogram when called), or you can provide a string and a list of variables and the common HistFactory class will take care of things for you.  Eg:

```python
from cmsl1t.hist.hist_collection import Binning
pileup = Binning([0, 10, 15, 20, 30, 999], "pileup")               
coll = hist.HistogramCollection([pileup], "Hist1D", 10, 0, 5)
coll[13].fill(1) 
```

I've also:
 - written most of the unit tests I think we need for the new bits of code
 - Rearranged the histograms code, creating a `cmsl1t/hist` package.  After the next stages we can delete the collections package I think.

Next steps are to:
 - Get a first working "plotter", eg for the Turnon curves, that uses the new histogram collection class
 - Rewrite some of the analyzers to use only these newer features

Do you think we should already merge this request into master and then continue work on the above two points on a new PR?  I'd be in favour of that approach, I think.

**Pull request merged**